packages/sdk/app-framework/src/plugins/plugin-surface/SurfaceContext.ts

```diff
@@ -0,0 +1,108 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { createContext, useContext, type JSX, type ForwardedRef, type PropsWithChildren, type ReactNode } from 'react';
+
+import { raise } from '@dxos/debug';
+import { type GuardedType, type MakeOptional } from '@dxos/util';
+
+import { type ErrorBoundary } from './ErrorBoundary';
+
+/**
+ * SurfaceProps are the props that are passed to the Surface component.
+ */
+export type SurfaceProps<T extends Record<string, any> = Record<string, unknown>> = PropsWithChildren<{
+  /**
+   * If specified, the Surface will be wrapped in an error boundary.
+   * The fallback component will be rendered if an error occurs.
+   */
+  fallback?: ErrorBoundary['props']['fallback'];
+
+  /**
+   * If specified, the Surface will be wrapped in a suspense boundary.
+   * The placeholder component will be rendered while the surface component is loading.
+   */
+  placeholder?: ReactNode;
+}> &
+  MakeOptional<CoreSurfaceProps<T>, 'id' | 'data'> & {
+    /**
+     * Additional props to pass to the component.
+     * These props are not used by Surface itself but may be used by components which resolve the surface.
+     */
+    [key: string]: unknown;
+  };
+
+// NOTE: If `[key: string]: unknown` is included in shared types, when re-used other fields become unknown as well.
+type CoreSurfaceProps<T extends Record<string, any> = Record<string, unknown>> = {
+  /**
+   * ID for debugging.
+   */
+  id: string;
+
+  /**
+   * Role defines how the data should be rendered.
+   */
+  role: string;
+
+  /**
+   * The data to be rendered by the surface.
+   */
+  data: T;
+
+  /**
+   * If more than one component is resolved, the limit determines how many are rendered.
+   */
+  limit?: number | undefined;
+};
+
+type SurfaceComponentProps<T extends Record<string, any> = Record<string, unknown>> = PropsWithChildren<
+  CoreSurfaceProps<T> & { [key: string]: unknown }
+>;
+
+/**
+ * React component used to render a surface once is has matched.
+ */
+export type SurfaceComponent<T extends Record<string, any> = Record<string, unknown>> = (
+  props: SurfaceComponentProps<T>,
+  forwardedRef: ForwardedRef<HTMLElement>,
+) => JSX.Element | null;
+
+/**
+ * Determines the priority of the surface when multiple components are resolved.
+ */
+export type SurfaceDisposition = 'default' | 'hoist' | 'fallback';
```

Comment:
define hoist
remove 'default' as an option? just used undefined as default?

Reply:
I think there’s value in explicitly making a positive space for a behavior that to us seems “default” in situations like these, it may need layers of override where `disposition: undefined` looks weird or we may want to rename it and the “default” behavior becomes something else.

Reply:
To that end, I think the term ‘default’ should be replaced by something that describes more what it does than its current utility. Maybe ‘static‘ or ‘normal‘?

Reply:
yeah it should be easy to change because it's only used internally currently. It could just as easily be undefined but I thought having it be a string value might be clearer for future debugging tools